### PR TITLE
Suport modules with multiple tasks

### PR DESCRIPTION
**Pull request by matthewpwilson**
refs #246 

**What this PR does / why we need it**:
Today Caikit allows only a single task to be associated with a module. However, some ML models may support multiple tasks. For example, a multi-modal model may support inference tasks for each data modality. The only way to achieve this currently is to create two modules, and a model that refers to each of them. This means the model weights may be loaded twice, wasting memory. 

This PR implements multi-task support as follows:

- A new optional `tasks` parameter on `@module` allows you to specify a list of tasks. This takes precedence over `task`.
- Inference methods for each task are annotated with `@TaskClass.taskmethod()`.
- As a consequence of this, some of the task validation is relaxed as it's now valid to have multiple tasks specified in the class hierarchy of your module. 
- Services are generated for each task.
- At inference time, the global predict servicer looks up the correct service based on the task.
 
**Special notes for your reviewer**:
There are a few other possible considerations I haven't addressed, happy to investigate if we think they need further thought:
- It's now possible to have a module without a generic `run()` implementation.  Should ModuleBase (or maybe a new `MultiTaskModuleBase`) provide a default implementation of `run()` that attempts to delegate to one of the `taskmethod`s baesd on the parameters? Should we require a multi-task module still implement a `run()`?
- What does training mean for multi-task modules? 
- I've introduced `module.TASK_CLASSES` which contains all the tasks implemented by the module. I've kept `TASK_CLASS` for compatibility, but is this necessary? 
- Any other impacts I've missed?

**If applicable**:
- [ ] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: caikit/runtime/http_server/http_server.py
Comment (joerunde) [inline caikit/runtime/http_server/http_server.py:371]: q!!!!

Comment (matthewpwilson) [reply caikit/runtime/http_server/http_server.py]: 🤦 fixed!

**Comment by joerunde** [comment]
Thanks @matthewpwilson! This looks super great and is exactly where we want to take this moving forward.

I'll address some of your other considerations here:

> It's now possible to have a module without a generic run() implementation ...

I actually thought this was already the case, since you can define a module with a single `@Task.taskmethod` that's not named run. I could be mistaken about that, but regardless I don't think that's a huge deal and I don't think we need to require that modules override a `run()` method

> What does training mean for multi-task modules?

We discussed this with many of the caikit maintainers, we think we'd probably leave it up to the module authors to decide if a single `train` call always trains a model that can do all tasks, or if it may train a model that can only do some subset of the tasks that the module supports. This has some implications of how the trained models can be used, we figured the lightest-weight way to deal with this right now is to add a generic `TaskUnsupportedException` that a module's inference methods may raise if the loaded model cannot perform that task. A future improvement might be to add a `model.supported_tasks() -> Set[Task]` api that each module can optionally implement so that a caikit runtime can accurately report which loaded models can be used for which task apis, but that certainly doesn't have to happen here.

> I've introduced module.TASK_CLASSES which contains all the tasks implemented by the module. I've kept TASK_CLASS for compatibility, but is this necessary?

Right, it probably isn't. It'd be great if we can refactor this down to a list or set of tasks held internally on the module class.
There are also quite a few places where I see `module.TASK_CLASS`  referenced outside of decorator.py, so this would be a good forcing function to stick a proper `@property` on modules to expose a set of task classes (and maybe inference / training signatures too?). Would you mind doing that bit of refactoring and cleanup on this PR?

File: caikit/runtime/service_generation/create_service.py
Comment (joerunde) [inline caikit/runtime/service_generation/create_service.py:125]: oof, yeah this was always a kludge here to begin with. I'd much rather be accessing public properties of the module class when we're outside of the `caikit.core.modules` package.

Comment (joerunde) [reply caikit/runtime/service_generation/create_service.py:125]: I believe in python you can set class-level `@property`s by sticking them on the meta-class. So something like this in `caikit/core/modules/meta.py` should work:

```
class _ModuleBaseMeta(abc.ABCMeta):
    ...

    @property
    def tasks(cls) -> Set[TaskBase]:
        return cls._TASK_CLASSES
```

**Review by joerunde — CHANGES_REQUESTED** [review]
Looks fantastic!

If you can get that refactor in to just hold `TASK_CLASSES` and cleanup the usage outside the core/modules package this should be good to merge

File: caikit/runtime/service_generation/create_service.py
Comment (matthewpwilson) [reply caikit/runtime/service_generation/create_service.py]: I've implemented this property and also `ModuleBase.get_inference_signatures` to replace this use of `_TASK_INFERENCE_SIGNATURES`.

**Comment by matthewpwilson** [comment]
> If you can get that refactor in to just hold `TASK_CLASSES` and cleanup the usage outside the core/modules package this should be good to merge

Getting rid of references to `TASK_CLASS` threw up a few other places that needed to consdier multiple tasks in a module.  In particular I had to refactor the handling of included/excluded tasks for service generation quite a bit, please let me know what you think.

**Comment by matthewpwilson** [comment]
A new test was introduced that required changes due to the removal of TASK_CLASS, I've fixed it.

File: caikit/runtime/service_factory.py
Comment (joerunde) [inline caikit/runtime/service_factory.py:290]: Ah, I think I skimmed past this on first review and now I understand your question in the PR body better. I don't have any great ideas on how to name the training rpc now that a module may have multiple tasks. We'll probably need to address that soon-ish, but this is fine to go with for now.

This also highlights the extra coupling between this logic and the rpc implementation- any change in the naming scheme over in `rpc.py` has to be reflected here.
There's probably a better way to get at the reference to the `ModuleClassTrainRPC` reference and return the request class from that, but that doesn't need to be fixed in this PR.

File: caikit/runtime/service_generation/rpcs.py
Comment (joerunde) [inline caikit/runtime/service_generation/rpcs.py:156]: 🌶️

File: caikit/runtime/servicers/global_predict_servicer.py
Comment (joerunde) [inline caikit/runtime/servicers/global_predict_servicer.py:353]: very pythonic 👍

**Review by joerunde — APPROVED** [review]
Ship it!

We'll follow up with some design discussion about the training API but we don't want to hold this change up on it

**Comment by markstur** [comment]
WOO HOO!  :tada:   Thanks!

**Pull request merged**
